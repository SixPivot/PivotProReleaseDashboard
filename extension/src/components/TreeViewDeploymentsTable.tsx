import * as React from 'react'
import { ITreeColumn, renderExpandableTreeCell, Tree } from 'azure-devops-ui/TreeEx'
import { ITreeItem, ITreeItemEx, TreeItemProvider } from 'azure-devops-ui/Utilities/TreeItemProvider'
import { IItemProvider } from 'azure-devops-ui/Utilities/Provider'
import { IReadonlyObservableValue } from 'azure-devops-ui/Core/Observable'
import { IDeploymentTableItem, IPipelineInstance, IDashboardEnvironmentColumn } from '../types'
import { IEnvironmentInstance } from '../types'
import { SimpleTableCell } from 'azure-devops-ui/Table'
import { useState, useEffect } from 'react'
import { ITreeItemProvider } from 'azure-devops-ui/Utilities/TreeItemProvider'
import { DeploymentTableCell } from './DeploymentTableCell'
import { getBuild, getBuildTimeline } from '../api/BuildClient'
import { getPipelineApprovals } from '../api/PipelineApprovalsClient'

export const TreeViewDeploymentsTable = (props: {
    environments: IEnvironmentInstance[]
    pipelines: IPipelineInstance[]
    projectName?: string
}): JSX.Element => {
    const { environments, pipelines, projectName } = props
    const [folderViewItemProvider, setFolderViewItemProvider] = useState<ITreeItemProvider<IDeploymentTableItem>>()

    // State to hold build names and approval names for each environment instance
    const [buildNames, setBuildNames] = useState<Record<string, string>>({})
    const [approvalNames, setApprovalNames] = useState<Record<string, string>>({})

    useEffect(() => {
        if (pipelines && environments) buildTreeView()
    }, [pipelines, environments, buildNames, approvalNames])

    useEffect(() => {
        if (!projectName) return

        const newBuildNames: Record<string, string> = {}
        const newApprovalNames: Record<string, string> = {}

        // Function to fetch build names and approval information
        async function fetchBuildNames() {
            const pending: Array<Promise<void>> = []
            let totalKeys = 0
            pipelines.forEach((pipeline) => {
                Object.keys(pipeline.environments).forEach((envName: string) => {
                    const deploymentInstance = pipeline.environments[envName]
                    const key = pipeline.key + ':' + envName
                    totalKeys++
                    pending.push(
                        (async () => {
                            const build = await getBuild(projectName!, deploymentInstance.buildId!)
                            newBuildNames[key] = build?.buildNumber || deploymentInstance.value

                            if (!build) return

                            const buildId = build.id

                            const timeline = await getBuildTimeline(projectName!, deploymentInstance.buildId!)

                            if (!timeline) {
                                console.warn(`No timeline found for build ID: ${buildId}`)
                                return
                            }

                            // See if there are any Checkpoint.Approval records in the timeline
                            // whose parent's parent is type 'Stage' and has the same name as the stageName
                            const approvalTimelineRecord = timeline.records.find((record) => record.type === 'Checkpoint.Approval')

                            if (!approvalTimelineRecord) {
                                console.warn(`No approval timeline record found for build ID: ${buildId}`)
                                return
                            }

                            const checkpointRecord = timeline.records.find((record) => record.id === approvalTimelineRecord.parentId)
                            if (!checkpointRecord) {
                                console.warn(`No checkpoint record found for approval timeline record ID: ${approvalTimelineRecord.id}`)
                                return
                            }

                            const stageRecord = timeline.records.find(
                                (record) =>
                                    record.id === checkpointRecord.parentId &&
                                    record.type === 'Stage' &&
                                    record.name === deploymentInstance.stageName
                            )
                            if (!stageRecord) {
                                // This is expected if there is no approval required for this stage
                                return
                            }

                            const approvals = await getPipelineApprovals(projectName!)

                            if (!approvals || approvals.length === 0) {
                                console.warn(`No approvals found for project: ${projectName}`)
                                return
                            }

                            const approval = approvals.find((a) => a.id === approvalTimelineRecord.id)

                            if (!approval) {
                                console.warn(`No approval found for timeline record ID: ${approvalTimelineRecord.id}`)
                                return
                            }

                            const approver = approval?.steps[0]?.actualApprover?.displayName

                            if (approver) {
                                newApprovalNames[key] = approver
                            } else {
                                console.warn(`No approver found for approval ID: ${approval.id}`)
                            }
                        })()
                    )
                })
            })
            await Promise.all(pending)
            setBuildNames(newBuildNames)
            setApprovalNames(newApprovalNames)
        }
        fetchBuildNames()
    }, [pipelines, projectName])

    const buildTreeView = () => {
        let treeNodeItems: ITreeItem<IDeploymentTableItem>[] = []
        pipelines!.forEach((pipeline: IPipelineInstance) => {
            let folder = pipeline.environments[Object.keys(pipeline.environments)[0]]?.folder!
            let paths = folder.split('\\')

            // Remove the first empty path, set paths to [] at root level
            paths.shift()
            if (paths.length === 1 && paths[0] === '') paths = []

            addToTreeNodes(treeNodeItems, paths, pipeline)
        })

        setFolderViewItemProvider(new TreeItemProvider<IDeploymentTableItem>(treeNodeItems))
    }

    const addToTreeNodes = (pathChildren: ITreeItem<IDeploymentTableItem>[], paths: string[], pipelineInfo: IPipelineInstance) => {
        if (paths.length === 0) {
            pathChildren.push({
                data: {
                    name: pipelineInfo.name,
                    pipeline: pipelineInfo,
                } as IDeploymentTableItem,
                childItems: [],
                expanded: true,
            })
        } else {
            let existingNode = pathChildren.find((item) => item.data.name === paths[0])

            if (!existingNode) {
                existingNode = {
                    data: {
                        name: paths[0],
                    } as IDeploymentTableItem,
                    childItems: [],
                    expanded: true,
                }

                pathChildren.push(existingNode)
            }

            paths.shift()
            addToTreeNodes(existingNode!.childItems!, paths, pipelineInfo)
        }
    }

    const getFolderViewColumns = (): ITreeColumn<IDeploymentTableItem>[] => {
        let columns: ITreeColumn<IDeploymentTableItem>[] = []

        columns.push({
            id: 'name',
            name: '',
            width: 300,
            renderCell: renderExpandableTreeCell,
        })

        let dynamicColumns = environments.map((env) => {
            return {
                id: env.name!,
                name: env.name,
                width: !env.name ? 300 : 200,
                renderCell: renderTreeViewCell,
                isFixedColumn: true,
            }
        })

        return columns.concat(dynamicColumns)
    }

    const renderTreeViewCell = <T extends IDeploymentTableItem>(
        _rowIndex: number,
        columnIndex: number,
        treeColumn: ITreeColumn<T>,
        treeItem: ITreeItemEx<T>,
        _ariaRowIndex?: number | undefined,
        _role?: string
    ): JSX.Element => {
        let pipeline = treeItem.underlyingItem.data.pipeline

        // If row isn't a leaf node, then return no data indicator
        if (treeItem.underlyingItem.childItems && treeItem.underlyingItem.childItems.length > 0) {
            return (
                <SimpleTableCell key={'col-' + columnIndex} columnIndex={columnIndex}>
                    <div className="no-data">-</div>
                </SimpleTableCell>
            )
        }

        // If there's no pipeline (folder node) or no environment data for this pipeline
        if (!pipeline || !pipeline.environments[treeColumn!.name!]) {
            return (
                <SimpleTableCell key={'col-' + columnIndex} columnIndex={columnIndex}>
                    <div className="no-data">-</div>
                </SimpleTableCell>
            )
        }

        // Get build name and approval name for this pipeline/environment combination
        let buildName: string | undefined = undefined
        let approvalName: string | undefined = undefined
        const env = pipeline.environments[treeColumn!.name!]
        if (env) {
            const key = pipeline.key + ':' + treeColumn!.name!
            buildName = buildNames[key] || env.value
            approvalName = approvalNames[key]

            // Debug: log when we have enhanced data
            if (buildNames[key] && buildNames[key] !== env.value) {
                console.debug('TreeView using enhanced build name:', buildNames[key], 'for key:', key)
            }

            if (approvalNames[key]) {
                console.debug('TreeView using approval name:', approvalNames[key], 'for key:', key)
            }
        }

        // Create a compatible table column for DeploymentTableCell
        const tableColumn: IDashboardEnvironmentColumn = {
            id: treeColumn.id,
            name: treeColumn.name,
            minWidth: 100,
            maxWidth: typeof treeColumn.width === 'number' ? treeColumn.width : 200,
            width: typeof treeColumn.width === 'number' ? treeColumn.width : 200,
            renderCell: () => <></>,
        }

        return (
            <DeploymentTableCell
                columnIndex={columnIndex}
                tableColumn={tableColumn}
                key={'col-' + columnIndex}
                tableItem={pipeline}
                buildName={buildName}
                approvalName={approvalName}
            />
        )
    }

    return (
        <>
            {folderViewItemProvider && (
                <Tree<IDeploymentTableItem>
                    className="deployments-table"
                    columns={getFolderViewColumns()}
                    itemProvider={
                        folderViewItemProvider as IItemProvider<
                            ITreeItemEx<IDeploymentTableItem> | IReadonlyObservableValue<ITreeItemEx<IDeploymentTableItem>>
                        >
                    }
                    onToggle={(_, treeItem: ITreeItemEx<IDeploymentTableItem>) => {
                        folderViewItemProvider!.toggle(treeItem.underlyingItem)
                    }}
                    scrollable={true}
                />
            )}
        </>
    )
}
