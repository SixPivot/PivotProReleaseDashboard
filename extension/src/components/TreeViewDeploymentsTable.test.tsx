// Mock SafeAgo to avoid time-related test flakiness
jest.mock('./SafeAgo', () => ({
    SafeAgo: ({ date }: { date: Date }) => <span data-testid="safe-ago">{date.toISOString()}</span>,
}))

// Mock the API calls
jest.mock('../api/BuildClient', () => ({
    getBuild: jest.fn().mockResolvedValue(null),
    getBuildTimeline: jest.fn().mockResolvedValue(null),
}))

jest.mock('../api/PipelineApprovalsClient', () => ({
    getPipelineApprovals: jest.fn().mockResolvedValue([]),
}))

/// <reference types="node" />
const OriginalDateTimeFormat = Intl.DateTimeFormat
jest.spyOn(Intl, 'DateTimeFormat').mockImplementation((...args) => {
    return new OriginalDateTimeFormat(['en-US'], { ...args[1], timeZone: 'UTC' })
})

import * as React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { TreeViewDeploymentsTable } from './TreeViewDeploymentsTable'
import { IEnvironmentInstance, IPipelineInstance } from '../types'

const TaskResult = { Succeeded: 2, Failed: 8 }

const mockEnvironments: IEnvironmentInstance[] = [
    {
        name: 'Development',
    },
    {
        name: 'Production',
    },
]

const mockPipelines: IPipelineInstance[] = [
    {
        key: 'pipeline-1',
        name: 'Pipeline 1',
        uri: 'https://dev.azure.com/org/project/_build/results?buildId=1',
        environments: {
            Development: {
                value: 'Build 123',
                buildId: 1,
                uri: 'https://dev.azure.com/org/project/_build/results?buildId=1',
                result: TaskResult.Succeeded,
                finishTime: new Date('2025-06-29T12:00:00Z'),
                environmentId: 1,
                stageName: 'Dev',
                folder: '\\Frontend',
            },
            Production: {
                value: 'Build 123',
                buildId: 1,
                uri: 'https://dev.azure.com/org/project/_build/results?buildId=1',
                result: TaskResult.Failed,
                finishTime: new Date('2025-06-29T14:00:00Z'),
                environmentId: 2,
                stageName: 'Prod',
                folder: '\\Frontend',
            },
        },
    },
    {
        key: 'pipeline-2',
        name: 'Pipeline 2',
        uri: 'https://dev.azure.com/org/project/_build/results?buildId=2',
        environments: {
            Development: {
                value: 'Build 456',
                buildId: 2,
                uri: 'https://dev.azure.com/org/project/_build/results?buildId=2',
                result: TaskResult.Succeeded,
                finishTime: new Date('2025-06-29T10:00:00Z'),
                environmentId: 1,
                stageName: 'Dev',
                folder: '\\Backend\\API',
            },
        },
    },
]

describe('TreeViewDeploymentsTable', () => {
    beforeEach(() => {
        // Clear all mocks before each test to prevent interference
        jest.clearAllMocks()
    })

    it('renders empty table when no pipelines are provided', () => {
        // Don't provide projectName to avoid triggering async effects
        render(<TreeViewDeploymentsTable environments={[]} pipelines={[]} />)
        expect(document.querySelector('.deployments-table')).toBeInTheDocument()
        // Should have only header row, no data rows
        expect(document.querySelectorAll('.bolt-tree-row').length).toBe(0)
    })

    it('renders the tree view with correct structure after data loads', async () => {
        // Don't provide projectName to avoid async API calls in this structural test
        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} />)

        await waitFor(() => {
            expect(document.querySelector('.deployments-table')).toBeInTheDocument()
        })

        // Check for folder structure
        expect(screen.getByText('Frontend')).toBeInTheDocument()
        expect(screen.getByText('Backend')).toBeInTheDocument()
        expect(screen.getByText('API')).toBeInTheDocument()
    })

    it('renders pipeline names as leaf nodes', async () => {
        // Don't provide projectName to avoid async API calls in this structural test
        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} />)

        await waitFor(() => {
            expect(screen.getByText('Pipeline 1')).toBeInTheDocument()
            expect(screen.getByText('Pipeline 2')).toBeInTheDocument()
        })
    })

    it('renders deployment status and build links for each environment', async () => {
        // Don't provide projectName to avoid async API calls, focus on basic rendering
        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} />)

        await waitFor(() => {
            // Check for build links - there can be multiple for the same build across different environments
            const build123Links = screen.getAllByRole('link', { name: /Build 123/i })
            expect(build123Links.length).toBeGreaterThan(0)

            const build456Links = screen.getAllByRole('link', { name: /Build 456/i })
            expect(build456Links.length).toBeGreaterThan(0)
        })
    })

    it('renders finish times using SafeAgo component', async () => {
        // Don't provide projectName to avoid async API calls
        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} />)

        await waitFor(() => {
            const safeAgoElements = screen.getAllByTestId('safe-ago')
            expect(safeAgoElements.length).toBeGreaterThan(0)
            expect(safeAgoElements[0]).toHaveTextContent('2025-06-29T12:00:00.000Z')
        })
    })

    it('shows dash for non-leaf nodes in environment columns', async () => {
        // Don't provide projectName to avoid async API calls
        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} />)

        await waitFor(() => {
            const dashElements = screen.getAllByText('-')
            expect(dashElements.length).toBeGreaterThan(0)
        })
    })

    it('handles pipelines with no folder (root level)', async () => {
        const rootPipelines: IPipelineInstance[] = [
            {
                key: 'root-pipeline',
                name: 'Root Pipeline',
                uri: 'https://dev.azure.com/org/project/_build/results?buildId=3',
                environments: {
                    Development: {
                        value: 'Build 789',
                        buildId: 3,
                        uri: 'https://dev.azure.com/org/project/_build/results?buildId=3',
                        result: TaskResult.Succeeded,
                        finishTime: new Date('2025-06-29T16:00:00Z'),
                        environmentId: 1,
                        stageName: 'Dev',
                        folder: '\\',
                    },
                },
            },
        ]

        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={rootPipelines} />)

        await waitFor(() => {
            expect(screen.getByText('Root Pipeline')).toBeInTheDocument()
        })
    })

    it('matches snapshot', async () => {
        // Don't provide projectName to avoid async API calls affecting snapshot
        const { container } = render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} />)

        await waitFor(() => {
            expect(document.querySelector('.deployments-table')).toBeInTheDocument()
        })

        expect(container).toMatchSnapshot()
    })

    it('accepts projectName prop without errors', async () => {
        // This test simply ensures that providing projectName doesn't break the component
        // We don't assert on API calls to avoid timing issues
        render(<TreeViewDeploymentsTable environments={mockEnvironments} pipelines={mockPipelines} projectName="TestProject" />)

        await waitFor(
            () => {
                expect(document.querySelector('.deployments-table')).toBeInTheDocument()
            },
            { timeout: 3000 }
        )

        // Just verify the component rendered successfully with the projectName prop
        expect(screen.getByText('Pipeline 1')).toBeInTheDocument()
        expect(screen.getByText('Pipeline 2')).toBeInTheDocument()
    })
})
