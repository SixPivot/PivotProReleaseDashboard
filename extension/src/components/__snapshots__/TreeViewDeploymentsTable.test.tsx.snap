// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeViewDeploymentsTable matches snapshot 1`] = `
<div>
  <div
    class="bolt-table-container flex-grow v-scroll-auto"
  >
    <table
      aria-colcount="3"
      aria-rowcount="6"
      class="deployments-table bolt-table bolt-table-show-lines bolt-list body-m relative scroll-hidden"
      role="treegrid"
      style="width: 716px;"
      tabindex="0"
    >
      <colgroup>
        <col
          style="width: 8px;"
        />
        <col
          style="width: 300px;"
        />
        <col
          style="width: 200px;"
        />
        <col
          style="width: 200px;"
        />
        <col
          style="width: 8px;"
        />
      </colgroup>
      <thead>
        <tr
          aria-rowindex="1"
          class="bolt-table-header-row"
          data-row-index="-1"
          role="row"
        >
          <th
            aria-hidden="true"
            class="bolt-table-header-border"
            role="presentation"
          />
          <th
            aria-colindex="1"
            aria-label="Blank"
            aria-readonly="true"
            class="bolt-table-header-border"
          />
          <th
            aria-colindex="2"
            aria-label="Development"
            aria-readonly="true"
            class="bolt-table-header-cell col-header-1"
            data-column-index="1"
            data-focuszone="false"
            id="__bolt-th-col-2-Development"
            role="columnheader"
            tabindex="-1"
          >
            <div
              class="bolt-table-header-cell-content flex-row"
            >
              <div
                class="scroll-hidden"
                id="__bolt-th-col-content-12"
              >
                <div
                  class="bolt-table-header-cell-text text-ellipsis body-s"
                >
                  <span
                    class=""
                  >
                    Development
                  </span>
                </div>
              </div>
              <div />
            </div>
          </th>
          <th
            aria-colindex="3"
            aria-label="Production"
            aria-readonly="true"
            class="bolt-table-header-cell col-header-2"
            data-column-index="2"
            data-focuszone="false"
            id="__bolt-th-col-3-Production"
            role="columnheader"
            tabindex="-1"
          >
            <div
              class="bolt-table-header-cell-content flex-row"
            >
              <div
                class="scroll-hidden"
                id="__bolt-th-col-content-13"
              >
                <div
                  class="bolt-table-header-cell-text text-ellipsis body-s"
                >
                  <span
                    class=""
                  >
                    Production
                  </span>
                </div>
              </div>
              <div />
            </div>
          </th>
          <th
            aria-hidden="true"
            class="bolt-table-header-border"
            role="presentation"
          />
        </tr>
      </thead>
      <tbody
        class="relative"
      >
        <tr
          class="bolt-list-row-spacer invisible"
          role="presentation"
        >
          <td
            class="bolt-list-cell-spacer invisible"
            colspan="5"
            style="height: 0px;"
          />
        </tr>
        <tr
          class="bolt-list-row-spacer invisible"
          role="presentation"
        >
          <td
            class="bolt-list-cell-spacer invisible"
            colspan="5"
            style="height: 0px;"
          />
        </tr>
        <tr
          aria-busy="false"
          aria-expanded="true"
          aria-level="1"
          aria-rowindex="2"
          class="bolt-tree-row bolt-table-row bolt-list-row first-row"
          data-focuszone="focuszone-41"
          data-row-index="0"
          id="__bolt-0"
          role="row"
          tabindex="0"
        >
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
          <td
            aria-colindex="1"
            class="bolt-tree-cell bolt-table-cell bolt-list-cell"
            data-column-index="0"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <span
                class="fluent-icons-enabled"
              >
                <span
                  aria-hidden="true"
                  class="bolt-tree-expand-button font-size cursor-pointer flex-noshrink fabric-icon ms-Icon--ChevronDownMed small"
                  role="presentation"
                />
              </span>
              <span
                class="bolt-list-cell-child flex-row flex-center bolt-list-cell-text"
              >
                <span
                  class="text-ellipsis body-m"
                >
                  Frontend
                </span>
              </span>
            </div>
          </td>
          <td
            aria-colindex="2"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="1"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            aria-colindex="3"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="2"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
        </tr>
        <tr
          aria-busy="false"
          aria-expanded="true"
          aria-level="2"
          aria-rowindex="3"
          class="bolt-tree-row bolt-table-row bolt-list-row"
          data-focuszone="focuszone-41"
          data-row-index="1"
          id="__bolt-1"
          role="row"
          tabindex="-1"
        >
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
          <td
            aria-colindex="1"
            class="bolt-tree-cell bolt-table-cell bolt-list-cell"
            data-column-index="0"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <span
                class="fluent-icons-enabled"
              >
                <span
                  aria-hidden="true"
                  class="bolt-tree-expand-button font-size cursor-pointer invisible flex-noshrink fabric-icon ms-Icon--ChevronDownMed small"
                  role="presentation"
                  style="margin-left: 16px;"
                />
              </span>
              <span
                class="bolt-list-cell-child flex-row flex-center bolt-list-cell-text"
              >
                <span
                  class="text-ellipsis body-m"
                >
                  Pipeline 1
                </span>
              </span>
            </div>
          </td>
          <td
            aria-colindex="2"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="1"
            role="gridcell"
          >
            <div
              class="fontWeightSemiBold font-weight-semibold fontSizeM font-size-m bolt-table-cell-content-with-inline-link bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="flex-row flex-start"
              >
                <svg
                  aria-labelledby="__bolt-status-17-desc"
                  class="bolt-status flex-noshrink failed icon-large-margin status-icon"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    id="__bolt-status-17-desc"
                  >
                    Failed
                  </desc>
                  <circle
                    cx="8"
                    cy="8"
                    r="8"
                  />
                  <path
                    d="M10.984 5.004a.9.9 0 0 1 0 1.272L9.27 7.99l1.74 1.741a.9.9 0 1 1-1.272 1.273l-1.74-1.741-1.742 1.74a.9.9 0 1 1-1.272-1.272l1.74-1.74-1.713-1.714a.9.9 0 0 1 1.273-1.273l1.713 1.713 1.714-1.713a.9.9 0 0 1 1.273 0z"
                    fill="#fff"
                  />
                </svg>
                <div
                  class="flex-column wrap-text"
                >
                  <a
                    class="bolt-table-inline-link bolt-table-link no-underline-link bolt-link"
                    data-focuszone="focuszone-44"
                    href="https://dev.azure.com/org/project/_build/results?buildId=1"
                    rel="noopener"
                    tabindex="0"
                    target="_top"
                  >
                    Build 123
                  </a>
                  <div
                    class="finish-date"
                  >
                    <span
                      data-testid="safe-ago"
                    >
                      2025-06-29T12:00:00.000Z
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </td>
          <td
            aria-colindex="3"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="2"
            role="gridcell"
          >
            <div
              class="fontWeightSemiBold font-weight-semibold fontSizeM font-size-m bolt-table-cell-content-with-inline-link bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="flex-row flex-start"
              >
                <svg
                  aria-labelledby="__bolt-status-18-desc"
                  class="bolt-status flex-noshrink success icon-large-margin status-icon"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    id="__bolt-status-18-desc"
                  >
                    Success
                  </desc>
                  <circle
                    cx="8"
                    cy="8"
                    r="8"
                  />
                  <path
                    d="M6.062 11.144l-.003-.002-1.784-1.785A.937.937 0 1 1 5.6 8.031l1.125 1.124 3.88-3.88A.937.937 0 1 1 11.931 6.6l-4.54 4.54-.004.004a.938.938 0 0 1-1.325 0z"
                    fill="#fff"
                  />
                </svg>
                <div
                  class="flex-column wrap-text"
                >
                  <a
                    class="bolt-table-inline-link bolt-table-link no-underline-link bolt-link"
                    data-focuszone="focuszone-44"
                    href="https://dev.azure.com/org/project/_build/results?buildId=1"
                    rel="noopener"
                    tabindex="0"
                    target="_top"
                  >
                    Build 123
                  </a>
                  <div
                    class="finish-date"
                  >
                    <span
                      data-testid="safe-ago"
                    >
                      2025-06-29T14:00:00.000Z
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </td>
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
        </tr>
        <tr
          aria-busy="false"
          aria-expanded="true"
          aria-level="1"
          aria-rowindex="4"
          class="bolt-tree-row bolt-table-row bolt-list-row"
          data-focuszone="focuszone-41"
          data-row-index="2"
          id="__bolt-2"
          role="row"
          tabindex="-1"
        >
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
          <td
            aria-colindex="1"
            class="bolt-tree-cell bolt-table-cell bolt-list-cell"
            data-column-index="0"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <span
                class="fluent-icons-enabled"
              >
                <span
                  aria-hidden="true"
                  class="bolt-tree-expand-button font-size cursor-pointer flex-noshrink fabric-icon ms-Icon--ChevronDownMed small"
                  role="presentation"
                />
              </span>
              <span
                class="bolt-list-cell-child flex-row flex-center bolt-list-cell-text"
              >
                <span
                  class="text-ellipsis body-m"
                >
                  Backend
                </span>
              </span>
            </div>
          </td>
          <td
            aria-colindex="2"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="1"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            aria-colindex="3"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="2"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
        </tr>
        <tr
          aria-busy="false"
          aria-expanded="true"
          aria-level="2"
          aria-rowindex="5"
          class="bolt-tree-row bolt-table-row bolt-list-row"
          data-focuszone="focuszone-41"
          data-row-index="3"
          id="__bolt-3"
          role="row"
          tabindex="-1"
        >
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
          <td
            aria-colindex="1"
            class="bolt-tree-cell bolt-table-cell bolt-list-cell"
            data-column-index="0"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <span
                class="fluent-icons-enabled"
              >
                <span
                  aria-hidden="true"
                  class="bolt-tree-expand-button font-size cursor-pointer flex-noshrink fabric-icon ms-Icon--ChevronDownMed small"
                  role="presentation"
                  style="margin-left: 16px;"
                />
              </span>
              <span
                class="bolt-list-cell-child flex-row flex-center bolt-list-cell-text"
              >
                <span
                  class="text-ellipsis body-m"
                >
                  API
                </span>
              </span>
            </div>
          </td>
          <td
            aria-colindex="2"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="1"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            aria-colindex="3"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="2"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
        </tr>
        <tr
          aria-busy="false"
          aria-expanded="true"
          aria-level="3"
          aria-rowindex="6"
          class="bolt-tree-row bolt-table-row bolt-list-row"
          data-focuszone="focuszone-41"
          data-row-index="4"
          id="__bolt-4"
          role="row"
          tabindex="-1"
        >
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
          <td
            aria-colindex="1"
            class="bolt-tree-cell bolt-table-cell bolt-list-cell"
            data-column-index="0"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <span
                class="fluent-icons-enabled"
              >
                <span
                  aria-hidden="true"
                  class="bolt-tree-expand-button font-size cursor-pointer invisible flex-noshrink fabric-icon ms-Icon--ChevronDownMed small"
                  role="presentation"
                  style="margin-left: 32px;"
                />
              </span>
              <span
                class="bolt-list-cell-child flex-row flex-center bolt-list-cell-text"
              >
                <span
                  class="text-ellipsis body-m"
                >
                  Pipeline 2
                </span>
              </span>
            </div>
          </td>
          <td
            aria-colindex="2"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="1"
            role="gridcell"
          >
            <div
              class="fontWeightSemiBold font-weight-semibold fontSizeM font-size-m bolt-table-cell-content-with-inline-link bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="flex-row flex-start"
              >
                <svg
                  aria-labelledby="__bolt-status-19-desc"
                  class="bolt-status flex-noshrink failed icon-large-margin status-icon"
                  height="16"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    id="__bolt-status-19-desc"
                  >
                    Failed
                  </desc>
                  <circle
                    cx="8"
                    cy="8"
                    r="8"
                  />
                  <path
                    d="M10.984 5.004a.9.9 0 0 1 0 1.272L9.27 7.99l1.74 1.741a.9.9 0 1 1-1.272 1.273l-1.74-1.741-1.742 1.74a.9.9 0 1 1-1.272-1.272l1.74-1.74-1.713-1.714a.9.9 0 0 1 1.273-1.273l1.713 1.713 1.714-1.713a.9.9 0 0 1 1.273 0z"
                    fill="#fff"
                  />
                </svg>
                <div
                  class="flex-column wrap-text"
                >
                  <a
                    class="bolt-table-inline-link bolt-table-link no-underline-link bolt-link"
                    data-focuszone="focuszone-47"
                    href="https://dev.azure.com/org/project/_build/results?buildId=2"
                    rel="noopener"
                    tabindex="0"
                    target="_top"
                  >
                    Build 456
                  </a>
                  <div
                    class="finish-date"
                  >
                    <span
                      data-testid="safe-ago"
                    >
                      2025-06-29T10:00:00.000Z
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </td>
          <td
            aria-colindex="3"
            class="bolt-table-cell bolt-list-cell"
            data-column-index="2"
            role="gridcell"
          >
            <div
              class="bolt-table-cell-content flex-row flex-center"
            >
              <div
                class="no-data"
              >
                -
              </div>
            </div>
          </td>
          <td
            class="bolt-table-cell-compact bolt-table-cell bolt-list-cell"
            role="presentation"
          />
        </tr>
        <tr
          class="bolt-list-row-spacer invisible"
          role="presentation"
        >
          <td
            class="bolt-list-cell-spacer invisible"
            colspan="5"
            style="height: 0px;"
          />
        </tr>
        <tr
          class="bolt-list-row-spacer invisible"
          role="presentation"
        >
          <td
            class="bolt-list-cell-spacer invisible"
            colspan="5"
            style="height: 0px;"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
